File: 904. Fruit Into Baskets.py
class Solution:
    def totalFruit(self, tree: List[int]) -> int:
        result = 0
        f1, f2, f1count, f2count, f2concount = None, None, 0, 0, 0
        for fruit in tree:
            if fruit == f1:
                f1, f2, f1count, f2count, f2concount = f2, f1, f2count+f2concount, f1count, 1
            elif fruit == f2:
                f2concount += 1
            else:
                f1, f2, f1count, f2count, f2concount = f2, fruit, f2concount, 0, 1
            result = max(result, f1count+f2count+f2concount)
        return result

File: 904. Fruit Into Baskets.java
class Solution {
    public int totalFruit(int[] tree) {
        Integer b1=null, b2=null;
        int total=0, c1=0, c2=0, temp=0;
        for (int i : tree) {
            if (b1 == null) {
                b1 = i;
                temp++;
            } else if (b2 == null) {
                if (i == b1) {
                    temp++;
                } else {
                    b2 = b1;
                    c2 = temp;
                    b1 = i;
                    c1 = 0;
                    temp = 1;
                }
            } else {
                if (i == b1) {
                    temp++;
                } else if (i == b2) {
                    int t = b2;
                    b2 = b1;
                    b1 = t;
                    t = c2;
                    c2 = c1 + temp;
                    c1 = t;
                    temp = 1;
                } else {
                    b2 = b1;
                    c2 = temp;
                    b1 = i;
                    c1 = 0;
                    temp = 1;
                }
            }
            total = Math.max(total, c1 + c2 + temp);
        }
        return total;
    }
}